### PR TITLE
[docs] Create windows_debug.md

### PR DESCRIPTION
**Pull request by bobcao3**

### CODE REVIEW

File: docs/lang/articles/contribution/windows_debug.md [hunk 1]
```diff
@@ -0,0 +1,93 @@
+## Prerequisites:
+Able to build Taichi from source. (i.e. already have LLVM and related environment variables already setup)
```
Comment (neozhaoliang) [inline docs/lang/articles/contribution/windows_debug.md:2]: Add a link to the developer installation guide?

File: docs/lang/articles/contribution/windows_debug.md [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+1. Github Co-pilot
+2. VS Chromium for Code Search ( https://chromium.github.io/vs-chromium/ ). You might need to go to the release page to find the pre-release version needed for VS2022
+
+## Step 1. Turn on msbuild option in our build system:
+
+This is a new feature in setup.py (introduced in https://github.com/taichi-dev/taichi/pull/6724 ) that enables building Taichi with MSBuild & MSVC, and it generates a Visual Studio Project file to enable coding and seamless debugging through Visual Studio IDE.
+To turn it on, create a new environment variable (Type in windows search: environment variable, or set it temporarily with `$Env:` in PowerShell) called TAICHI_USE_MSBUILD, and set its value to 1 or ON.
+Now after cleaning up the _skbuild folder (in case the previously used generator is Ninja), run `python setup.py develop` and build Taichi from the source.
+
+## Step 2.  Navigate to the generated Visual Studio Project file
```
Comment (neozhaoliang) [inline docs/lang/articles/contribution/windows_debug.md:13]:
```suggestion
## Step 2. Navigate to the generated Visual Studio Project file
```

File: docs/lang/articles/contribution/windows_debug.md [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+![image13](https://user-images.githubusercontent.com/11663476/212577220-92a8a7cb-f6ff-4365-9808-0a7299be87cd.png)
+
+## Step 3. Navigating the source code of Taichi in Visual Studio
+
+The top right red box is the solution explorer, where you can see each submodule of Taichi grouped up by each “object file”, clicking on one will reveal something like this:
+
+![image10](https://user-images.githubusercontent.com/11663476/212577350-33912d21-0105-459b-8490-2aaee5c88ff6.png)
+![image11](https://user-images.githubusercontent.com/11663476/212577355-fec6837a-00fc-4f7b-8cdc-b369bc4bc015.png)
+
+Or if you prefer navigating the project using its folder structure, you can find the “Source Explorer” module, and it should reveal the normal Taichi source directory structure:
```
Comment (neozhaoliang) [inline docs/lang/articles/contribution/windows_debug.md:26]: Are the quotes Chinese characters?

File: docs/lang/articles/contribution/windows_debug.md [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+![image2](https://user-images.githubusercontent.com/11663476/212577382-4ff8e6de-e04b-4502-9dd9-7ebb75697693.png)
+
+Sometimes, you are quite familiar with the source code already and you know what you are trying to find, or when you are trying to figure out where a particular class is defined, using CodeSearch should be a quicker and easier way to find the source code. (Code Search is provided by the plugin VS Chromium, and you can find it through “View->Other Windows->Code Search”. I recommend keeping this window pinned somewhere so you always have access to it, it’s the greatest source navigation tool ever.
```
Comment (neozhaoliang) [inline docs/lang/articles/contribution/windows_debug.md:30]:
```suggestion
Sometimes, if you are familiar with the source code already and you know what you are trying to find, or if you are trying to figure out where a particular class is defined, using CodeSearch should be a quicker and easier way to find the source code. (Code Search is provided by the plugin VS Chromium, and you can find it through “View->Other Windows->Code Search”. I recommend keeping this window pinned somewhere so you always have access to it, it’s the greatest source navigation tool ever.
```

File: docs/lang/articles/contribution/windows_debug.md [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+
+(Currently Visual Studio has a bit of trouble parsing our `ir.h`, so ignore these errors related to classes and types it failed to find. Usually this goes away after a while.)
+
+## Step 4. Debugging Taichi
+
+To debug Taichi on Windows, you can use two methods. One is to launch a Taichi program from Visual Studio itself and use it like any other IDE, the other is to insert this statement in your Taichi program:
+`ti._lib.core.wait_for_debugger()`
+When this line of code is executed, the program will pause and now you can attach to this process from Visual Studio.
+To setup launching Taichi program from Visual Studio, go to “Debug->ALL_BUILD Debug Properties”, and click on the left side “Debugging”. This will reveal the debug launch settings. You should set “Command” to the python executable file you installed on your system (either in Conda or in other places). Then set the working directory to “..\..\..” where it points to the root of taichi source directory. Then you can set the argument to the python file you want to launch and its related options. Here is an example:
```
Comment (neozhaoliang) [inline docs/lang/articles/contribution/windows_debug.md:46]: The quotation marks?

File: docs/lang/articles/contribution/windows_debug.md (hunk 1 above)
Comment (Olinaaaloompa) [inline docs/lang/articles/contribution/windows_debug.md:2]:
```suggestion
You should be able to build Taichi from source and already have LLVM and related environment variables configured.
```

File: docs/lang/articles/contribution/windows_debug.md [hunk 6]
```diff
@@ -0,0 +1,93 @@
+## Prerequisites:
+Able to build Taichi from source. (i.e. already have LLVM and related environment variables already setup)
+Recommended plugins to install for Visual Studio:
```
Comment (Olinaaaloompa) [inline docs/lang/articles/contribution/windows_debug.md:3]:
```suggestion
Recommended Visual Studio plugins:
```

File: docs/lang/articles/contribution/windows_debug.md [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+1. Github Co-pilot
+2. VS Chromium for Code Search ( https://chromium.github.io/vs-chromium/ ). You might need to go to the release page to find the pre-release version needed for VS2022
```
Comment (Olinaaaloompa) [inline docs/lang/articles/contribution/windows_debug.md:5]:
```suggestion
2. [VS Chromium](https://chromium.github.io/vs-chromium/) for Code Search. You might need to go to the release page to find the pre-release version needed for VS2022
```

File: docs/lang/articles/contribution/windows_debug.md [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+
+## Step 1. Turn on msbuild option in our build system:
+
+This is a new feature in setup.py (introduced in https://github.com/taichi-dev/taichi/pull/6724 ) that enables building Taichi with MSBuild & MSVC, and it generates a Visual Studio Project file to enable coding and seamless debugging through Visual Studio IDE.
```
Comment (Olinaaaloompa) [inline docs/lang/articles/contribution/windows_debug.md:9]:
```suggestion
This is a new feature in setup.py, introduced in [PR #6724](https://github.com/taichi-dev/taichi/pull/6724), which enables building Taichi with MSBuild & MSVC. It generates a Visual Studio Project file to enable coding and seamless debugging through Visual Studio IDE.
```

File: docs/lang/articles/contribution/windows_debug.md [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+To turn it on, create a new environment variable (Type in windows search: environment variable, or set it temporarily with `$Env:` in PowerShell) called TAICHI_USE_MSBUILD, and set its value to 1 or ON.
+Now after cleaning up the _skbuild folder (in case the previously used generator is Ninja), run `python setup.py develop` and build Taichi from the source.
+
+## Step 2.  Navigate to the generated Visual Studio Project file
+Go to the build folder, in my case it is `_skbuild\win-amd64-3.9\cmake-build`, and double click on `taichi.sln` (Or open this SLN solution file from Visual Studio). This should open Visual Studio IDE with the setup for Taichi.
```
Comment (Olinaaaloompa) [inline docs/lang/articles/contribution/windows_debug.md:14]:
```suggestion
Go to the build folder, which is `_skbuild\win-amd64-3.9\cmake-build` in the author's case, and double click `taichi.sln`. You can also open this SLN solution file from Visual Studio. This should open Visual Studio IDE with the setup for Taichi.
```

File: docs/lang/articles/contribution/windows_debug.md [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+After Visual Studio launched, you should see something like this (layout might be different, you can reconfigure the layout to one that you like):
```
Comment (Olinaaaloompa) [inline docs/lang/articles/contribution/windows_debug.md:15]:
```suggestion
The following image shows a possible layout after Visual Studio is launched. The layout is open to reconfiguration.
```

**Review by Olinaaaloompa — CHANGES_REQUESTED** [review]

File: docs/lang/articles/contribution/windows_debug.md [hunk 11]
```diff
@@ -0,0 +1,93 @@
+## Prerequisites:
```
Comment (lin-hitonami) [inline docs/lang/articles/contribution/windows_debug.md:1]: We should add `sidebar_positions` and the title to let the doc site display this on the sidebar.
```suggestion
---
sidebar_positions: 9
---

# Debugging on Windows

## Prerequisites:
```

File: docs/lang/articles/contribution/windows_debug.md (hunk 3 above)
Comment (lin-hitonami) [reply docs/lang/articles/contribution/windows_debug.md:26]: No, they are not. English also has left and right quotation marks.

File: docs/lang/articles/contribution/windows_debug.md [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+
+![image10](https://user-images.githubusercontent.com/11663476/212577350-33912d21-0105-459b-8490-2aaee5c88ff6.png)
+![image11](https://user-images.githubusercontent.com/11663476/212577355-fec6837a-00fc-4f7b-8cdc-b369bc4bc015.png)
+
+Or if you prefer navigating the project using its folder structure, you can find the “Source Explorer” module, and it should reveal the normal Taichi source directory structure:
+
+![image2](https://user-images.githubusercontent.com/11663476/212577382-4ff8e6de-e04b-4502-9dd9-7ebb75697693.png)
+
+Sometimes, you are quite familiar with the source code already and you know what you are trying to find, or when you are trying to figure out where a particular class is defined, using CodeSearch should be a quicker and easier way to find the source code. (Code Search is provided by the plugin VS Chromium, and you can find it through “View->Other Windows->Code Search”. I recommend keeping this window pinned somewhere so you always have access to it, it’s the greatest source navigation tool ever.
+Say we are trying to pin down where we are defining all the pointer statements, just type in CodeSearch:
```
Comment (Olinaaaloompa) [inline docs/lang/articles/contribution/windows_debug.md:31]: For example, the following image shows how to pin down where all pointer statements are defined:

File: docs/lang/articles/contribution/windows_debug.md [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+
+![image16](https://user-images.githubusercontent.com/11663476/212577411-61c8ffd9-6b63-4eb9-a38b-b1b4f2a640dc.png)
+
+You can see here, by default CodeSearch uses the Windows search path wild cards with . ? and *. Here we matched all the pointer statement definitions by using `class *ptrstmt`. Clicking on each one will bring you over to that line in that file:
```
Comment (Olinaaaloompa) [inline docs/lang/articles/contribution/windows_debug.md:35]:
```suggestion
By default, CodeSearch uses the Windows search path wildcards with . ? and *. Here, we match all pointer statement definitions by using `class *ptrstmt`. Clicking on one of the results brings you over to the particular line in that file:
```

File: docs/lang/articles/contribution/windows_debug.md [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+
+![image8](https://user-images.githubusercontent.com/11663476/212577439-6cd6e888-fbd9-48c8-9a81-81cca3d9359f.png)
+
+(Currently Visual Studio has a bit of trouble parsing our `ir.h`, so ignore these errors related to classes and types it failed to find. Usually this goes away after a while.)
+
+## Step 4. Debugging Taichi
```
Comment (Olinaaaloompa) [inline docs/lang/articles/contribution/windows_debug.md:41]:
```suggestion
## Step 4. Debug Taichi
```

File: docs/lang/articles/contribution/windows_debug.md [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+Say we are trying to pin down where we are defining all the pointer statements, just type in CodeSearch:
+
+![image16](https://user-images.githubusercontent.com/11663476/212577411-61c8ffd9-6b63-4eb9-a38b-b1b4f2a640dc.png)
+
+You can see here, by default CodeSearch uses the Windows search path wild cards with . ? and *. Here we matched all the pointer statement definitions by using `class *ptrstmt`. Clicking on each one will bring you over to that line in that file:
+
+![image8](https://user-images.githubusercontent.com/11663476/212577439-6cd6e888-fbd9-48c8-9a81-81cca3d9359f.png)
+
+(Currently Visual Studio has a bit of trouble parsing our `ir.h`, so ignore these errors related to classes and types it failed to find. Usually this goes away after a while.)
```
Comment (Olinaaaloompa) [inline docs/lang/articles/contribution/windows_debug.md:39]:
```suggestion
:::note

Currently, Visual Studio has a bit of trouble parsing `ir.h`. You can ignore the errors reporting a failure to find the classes or types. Usually, they go away after a while.
```

File: docs/lang/articles/contribution/windows_debug.md [hunk 16]
```diff
@@ -0,0 +1,93 @@
...
+
+## Step 4. Debugging Taichi
+
+To debug Taichi on Windows, you can use two methods. One is to launch a Taichi program from Visual Studio itself and use it like any other IDE, the other is to insert this statement in your Taichi program:
```
Comment (Olinaaaloompa) [inline docs/lang/articles/contribution/windows_debug.md:43]:
```suggestion
To debug Taichi on Windows, you can use one of the following two methods. 

- Launch a Taichi program from Visual Studio and use it like any other IDE
- Insert `ti._lib.core.wait_for_debugger()` in your Taichi program to pause the program. You can then attach to this process from Visual Studio.
```

File: docs/lang/articles/contribution/windows_debug.md [hunk 17]
```diff
@@ -0,0 +1,93 @@
...
+`ti._lib.core.wait_for_debugger()`
```
Comment (Olinaaaloompa) [inline docs/lang/articles/contribution/windows_debug.md:44]:
```suggestion

```

File: docs/lang/articles/contribution/windows_debug.md (hunk 5 above)
Comment (Olinaaaloompa) [inline docs/lang/articles/contribution/windows_debug.md:46]:
```suggestion
To launch a Taichi program from Visual Studio, go to**Debug > ALL_BUILD Debug Properties** and click **Debugging** on the left side for the debug launch settings. Set “Command” to the python executable file installed on your system (either in Conda or in other places). Then, set the working directory to **..\..\..**, which points to the root of the Taichi source directory. Finally, set the argument to the python file you want to launch and its related options. For example:
```

File: docs/lang/articles/contribution/windows_debug.md [hunk 18]
```diff
@@ -0,0 +1,93 @@
...
+When this line of code is executed, the program will pause and now you can attach to this process from Visual Studio.
+To setup launching Taichi program from Visual Studio, go to “Debug->ALL_BUILD Debug Properties”, and click on the left side “Debugging”. This will reveal the debug launch settings. You should set “Command” to the python executable file you installed on your system (either in Conda or in other places). Then set the working directory to “..\..\..” where it points to the root of taichi source directory. Then you can set the argument to the python file you want to launch and its related options. Here is an example:
+
+![image6](https://user-images.githubusercontent.com/11663476/212577472-49959479-e0f5-4f7c-87c0-8b16fb53c07b.png)
+
+Hit OK or Apply to save this config, and click on the green Run button (says “local windows debugger”) to launch the program. Here you can see we hit a break point set in one of the optimization passes:
```
Comment (Olinaaaloompa) [inline docs/lang/articles/contribution/windows_debug.md:50]:
```suggestion
Hit OK or Apply to save this config, and click on the green Run button (says “local windows debugger”) to launch the program. Here, we hit a break point set in one of the optimization passes:
```

File: docs/lang/articles/contribution/windows_debug.md [hunk 19]
```diff
@@ -0,0 +1,93 @@
...
+
+![image5](https://user-images.githubusercontent.com/11663476/212577487-139cea4c-01ee-4589-89ff-f3daa2bdb982.png)
+
+If you use the other route and put a `wait_for_debugger()` in your python script, you can attach to the waiting process through “Debug->Attach to Process”. After you have attached, the program will resume automatically.
```
Comment (Olinaaaloompa) [inline docs/lang/articles/contribution/windows_debug.md:54]:
```suggestion
If you use the other route and put a `wait_for_debugger()` in your python script, you can attach to the waiting process through **Debug > Attach to Process**. Afterward, the program resumes automatically.
```

File: docs/lang/articles/contribution/windows_debug.md [hunk 20]
```diff
@@ -0,0 +1,93 @@
...
+
+## Step 5. (CPU) Performance profiling
+
+After clicking the debugging button (green run button), you might have noticed that a new window popped up called “Diagnostic Tools” that looks like this:
```
Comment (Olinaaaloompa) [inline docs/lang/articles/contribution/windows_debug.md:58]:
```suggestion
After the debugger (the green Run button) is enabled, a new window titled “Diagnostic Tools” pops up, as the following image shows:
```

**Review by Olinaaaloompa — CHANGES_REQUESTED** [review]

File: docs/lang/articles/contribution/windows_debug.md [hunk 21]
```diff
@@ -0,0 +1,93 @@
...
+Now if we hit pause on debugging, we can see the profiling window turns into this:
+
+![image7](https://user-images.githubusercontent.com/11663476/212577591-d593a3b4-a13b-47f7-ac25-a376f69fcb95.png)
+
+This is the collected profile of our program. We can select a region on the CPU timeline and only look at performance data in that region:
+
+![image14](https://user-images.githubusercontent.com/11663476/212577515-ebe3a000-8294-41c9-9355-73f6fe20837a.png)
+
+Here we can see the “Hot Path” of function calls. If we hit the “Open details” button, we can go into the detailed view containing a few views: “Caller/Callee”, “Call Tree”, “Modules”, “Functions”. Each of these presents performance data in a different way.
+In this particular program, we can see that the hot path is within Python, which we can’t do much to help. In this case, if we still want to optimize our part of code, we can go to the “Modules” view:
```
Comment (Olinaaaloompa) [inline docs/lang/articles/contribution/windows_debug.md:78]:
```suggestion
In this particular program, the hot path is within Python, and there is not much we can do to help. To optimize this part of code, go to the “Modules” view:  
```

File: docs/lang/articles/contribution/windows_debug.md [hunk 22]
```diff
@@ -0,0 +1,93 @@
...
+
+![image9](https://user-images.githubusercontent.com/11663476/212577614-9cb2dd9d-18c5-4900-a347-869f10f583e4.png)
+
+Where we can see that the `taichi_python` module (the C++ source code of Taichi) takes 65% of total CPU time, and then it splits its time down into the kernel and driver libraries etc. Here is the view when we expand our library module:
```
Comment (Olinaaaloompa) [inline docs/lang/articles/contribution/windows_debug.md:82]:
```suggestion
The `taichi_python` module (the C++ source code of Taichi) takes 65% of the total CPU time and splits its time down into the kernel and driver libraries. 

 Now, expand the library module:
```

File: docs/lang/articles/contribution/windows_debug.md [hunk 23]
```diff
@@ -0,0 +1,93 @@
...
+
+![image15](https://user-images.githubusercontent.com/11663476/212577640-87a0503c-72d8-4c4c-9306-1e4ee97e3796.png)
+
+Let’s say we want to focus on
```
Comment (Olinaaaloompa) [inline docs/lang/articles/contribution/windows_debug.md:86]:
```suggestion
Take this particular entry for example.
```

File: docs/lang/articles/contribution/windows_debug.md [hunk 24]
```diff
@@ -0,0 +1,93 @@
...
+
+![image4](https://user-images.githubusercontent.com/11663476/212577647-116bf750-54df-491b-8719-01e88ef526cd.png)
+
+We can right click on this line, and click “View in Call Tree”, and then click on “expand hot path”, we get something like this:
```
Comment (Olinaaaloompa) [inline docs/lang/articles/contribution/windows_debug.md:90]:
```suggestion
Right click on it and find **View in Call Tree > expand hot path**. The result is shown as below:
```

File: docs/lang/articles/contribution/windows_debug.md [hunk 25]
```diff
@@ -0,0 +1,93 @@
...
+
+![image1](https://user-images.githubusercontent.com/11663476/212577664-48f91acb-988a-463c-abe4-3f808d3159ad.png)
+
+Hmmmm, perhaps the Python stack is way too long and we don’t really care about it. Let’s scroll down to `taichi::lang::JITSessionCUDA::add_module`, and right click on “Set Root”, and now we get a much cleaner view where we see the function we care about is now the root of the call tree:
```
Comment (Olinaaaloompa) [inline docs/lang/articles/contribution/windows_debug.md:94]:
```suggestion
The Python stack is lengthy. We can scroll down to `taichi::lang::JITSessionCUDA::add_module` and right click on “Set Root” to get a much cleaner view. The function we care about is now the root of the call tree:
```

File: docs/lang/articles/contribution/windows_debug.md [hunk 26]
```diff
@@ -0,0 +1,93 @@
...
+
+![image12](https://user-images.githubusercontent.com/11663476/212577676-772d210b-11e8-4959-b573-28a73bbb47d9.png)
+
+Here we can see which pass takes the most amount of time, we can even see which line of our code corresponds to the most amount of time spent.
```
Comment (Olinaaaloompa) [inline docs/lang/articles/contribution/windows_debug.md:98]:
```suggestion
We can find out which pass and even which line of code are the most time-consuming.
```

File: docs/lang/articles/contribution/windows_debug.md [hunk 27]
```diff
@@ -0,0 +1,93 @@
...
+Granted in this case, most of the time is spent on LLVM here. We probably can’t do much for this specific case. However, this method applies everywhere and is extremely powerful and clear what performance problems are and gives us potential paths to fix these problems.
```
Comment (Olinaaaloompa) [inline docs/lang/articles/contribution/windows_debug.md:99]:
```suggestion
In this case, LLVM occupies most of the time and leaves limited room for optimization. However, this method applies everywhere and is most helpful to locate performance bottlenecks and facilitate problem-solving.
```

**Review by Olinaaaloompa — CHANGES_REQUESTED** [review]

File: docs/lang/articles/contribution/windows_debug.md [hunk 28]
```diff
@@ -0,0 +1,109 @@
...
+
+![image16](https://user-images.githubusercontent.com/11663476/212577411-61c8ffd9-6b63-4eb9-a38b-b1b4f2a640dc.png)
+
+By default, CodeSearch uses the Windows search path wildcards with . ? and *. For example, the following image shows how to pin down where all pointer statements are defined, and clicking on one of the results brings you over to the particular line in that file:
+
+![image8](https://user-images.githubusercontent.com/11663476/212577439-6cd6e888-fbd9-48c8-9a81-81cca3d9359f.png)
+
+:::note
+
+Currently, Visual Studio has a bit of trouble parsing `ir.h`. You can ignore the errors reporting a failure to find the classes or types. Usually, they go away after a while.
```
Comment (Olinaaaloompa) [inline docs/lang/articles/contribution/windows_debug.md:50]:
```suggestion
Currently, Visual Studio has a bit of trouble parsing `ir.h`. You can ignore the errors reporting a failure to find the classes or types. Usually, they go away after a while.

:::
```

**Review by Olinaaaloompa — CHANGES_REQUESTED** [review]

**Review by Olinaaaloompa — APPROVED** [review]
LGTM! Thanks!

**Pull request merged**